### ISSUE TITLE
How can i add the private dependency to podspec file?

### ISSUE THREAD

**Issue by nghiaphunguyen** [CocoaPods/CocoaPods#4921]
I try to add a dependency from private repo to podspec. But it doesn't work.

After run the command > pod repo push MyLib MyLib.podspec
[!] Invalid `MyLib.podspec` file: [!] Unsupported version requirements. Updating CocoaPods might fix the issue.

 #  from MyLib.podspec:16
 #  -------------------------------------------
 #    s.dependency 'ReachabilitySwift'
 >    s.dependency 'MyPrivateLib', :git => 'https://bitbucket.org/nghiaphunguyen/myprivatelib'
 #  -------------------------------------------

P/S: MyPrivateLib posted to the cocoapod via. https://guides.cocoapods.org/making/private-cocoapods.html

Nghia,
Thanks

**Comment by segiddins** [comment]
You can't specify source requirements in the podspec, only a version requirement.

**Comment by nghiaphunguyen** [comment]
@segiddins  i see but if i do that, how the podspec understand my private repo?

**Comment by segiddins** [comment]
Just add a normal dependency, the source used will come from the command line or the Podfile, depending on the command run.

**Comment by nghiaphunguyen** [comment]
what the format i should use to run the command > pod repo push MyLib MyLib.podspec to add another source?

**Comment by robipresotto** [comment]
the guy is asking about "private" pod. How we can add a private pod dependency into .podspec file?

**Comment by orta** [comment]
`pod repo push` allows you to define the private sources you want to refer to during the push - https://guides.cocoapods.org/terminal/commands.html#pod_repo_push

**Comment by iwasrobbed** [comment]
Adding a little more color to @orta's comment: https://guides.cocoapods.org/making/private-cocoapods.html

**Comment by patrickroteman** [comment]
@orta  How does using the pod repo push command work with an example project using the pod that is using relative file paths?

**Comment by zhonglaoban** [comment]
when you run pod lib lint you should config --sources=https://github.com/private-repo-name.git,and same when pod repo push ,like this pod lib lint name.podspec --sources=https://github.com/CocoaPods/Specs.git,192.168.0.100:Plutoy/Specs.git.here is the article http://www.pluto-y.com/cocoapod-private-pods-and-module-manager/

**Comment by patrickroteman** [comment]
Thanks!

**Comment by siddharthsuneel** [comment]
I am also having this same issue,
@nghiaphunguyen  did you any find the appropriate solution for this ??

**Comment by SebastianBoldt** [comment]
Our Problem is, that the there is a private pod with the same name as a spec inside the master-spec-repo. Sure, I can pass the sources via the command line, but this does not solve the issue if you got naming conflicts.

**Comment by rahulOYO** [comment]
Hi Team,
is there any update on this?

**Comment by MrBoog** [comment]
@rahulOYO  in that case, I guess we have to deploy our Podspecs to our private specs repo with `pod repo push REPO [NAME.podspec]` instead of specifying dependency source in Podspec.

https://stackoverflow.com/a/56581158/2007403

**Comment by konacurrents** [comment]
I just advanced my pod use to need a hierarchy of **private** pods. As stated above, there is no way to specify a private path in a podspec, only a version. The task of defining a path is pushed to the users of those pods, just like regular use of private pods. I feel that is a big burden once the private hierarchy gets complex.

My biggest issue is that if the pods were public, then the user wouldn't have to specify that pod relationship at all - it would just come with the pod install. Instead for private pod hierarchies, each podfile would need to know inside details of all the pods they use. That's a lot of complexity.

Obviously once a pod is uploaded to GitHub and made private or public this issues is simplified. The git lint tools would have to strip out a private path.

It sure seems this is a valid and important feature for CocoaPods to explore.
Thanks,

**Comment by Aure77** [comment]
@segiddins this issue should be re-opened or declared as feature request.

Why specifiying podspec is allowed from Podfile:
`pod "MyPrivateSDK", :podspec => "https://raw.githubusercontent.com/MyPrivateSDK/master/MyPrivateSDK.podspec"`

But not from from podspec dependency :
`s.dependency "MyPrivateSDK", :podspec => "https://raw.githubusercontent.com/MyPrivateSDK/master/MyPrivateSDK.podspec"`

How to deal with "non-public" dependencies in podspec file ?

**Comment by pacu** [comment]
This issue should be re-opened. There should be a way of including a dependency that is on another Specs repo.

**Comment by alimcharaniya** [comment]
Any movement here?

**Comment by konacurrents** [comment]
Sure seems to be a big gap in what is an awesome reuse capability. I too was just wondering the status as I really need this feature.

**Comment by orta** [comment]
I don't think this is something which will ever get built because the current answer is the right abstraction and we don't need to make the concepts leaky. It's likely that if you're hitting this - it's because you've probably not read up enough on the responsibilities on how Podfiles, Podspec repos and Podspecs systems work together.

If it's in your app: The podfile uses the ordering of sources to determine the podspec which gets picked up. If it's on the CLI (e.g. linting): there should be a flag to get the sources set up right for your case.

**Comment by konacurrents** [comment]
My understanding is a `Podfile` is for final apps (like examples) and `podspec` is to define the pod’s use of other pods.
So the common example for private pod developers is as follows:

1. I have a base type/class that needs to be shared amongst my private pods. (Call it `PodBaseType`)
2. This base type is a private pod on my file system. (Or in my private GitHub)
3. How do I reference this private pod from the dependencies (`podspec`) of my other private pods? (`PodA`, `PodB`). I know how to reference public pods from the `podspec`, but as the comments above, there is not way to reference other private pods.
4.  I know how to reference private pods from examples and end users of a private pod (using `Podfile`). Any internal code for `PodA` can use `PodBaseType` but only because the example included that pod. (This is a kludge)
5.  But anyone installing my `PodA` will also be required to include the `PodBaseType` in their `Podfile` or it won't compile. (Again a kludge as their example code doesn't necessarily reference `PodBaseType`) 

How is that leaky?

- If instead `PodBaseType` is public and installed in the _cocoapods_ cloud, then the `podspec` can reference it - and those grabbing my `PodA` will also get any pods referenced by that pod (but not from their example `Podfile`). 

I haven't created any public pods but have many private pods and I have to perform these kludges to get a large interrelated system to work. 

Please enlighten me on how to perform what I believe is a simple example that every object oriented reuse library requires?

Thanks

**Comment by orta** [comment]
Sounds to me like you need a private specs repo

**Comment by konacurrents** [comment]
@orta can you please elaborate on what a "private specs repo" is? Basically how can I achieve what was described
thanks.

**Comment by orta** [comment]
Please re-read this which is full of links about private spec repos, or consult stack overflow / duckduckgo. Our issues are not a support forum and I'm afraid I don't have the spare time for CocoaPods like I used to.

**Comment by c0d3runn3r** [comment]
@konacurrents You're not alone.  CocoaPods is a bit of a weird beast, both widely used but also feeling a bit tired.  And it's written in Ruby, which just seems weird to me.  _"Your mom is weird", someone is probably going to reply :)_  Anyway - there seems to be a lot less support around the podspecs creation workflow vs the more common "create a Podfile for my project" workflow.   **While it is possible to throw a git repo into a Podfile and have everything work fine, in order to lint your podspec all of your dependencies have to been in a private specs repo** (you can Google it) which means that someone has "pushed" it using ```pod repo push```.  This creates various beasties inside the repo that you don't get when you just ``git push``` your podspecs file.  Hope that helps.

**Comment by konacurrents** [comment]
I have had success with many private pods locally on my computer, and in my GitHub. But I cannot find a way to reference those in GitHub from other private pods (vs Podfile). I guess I can try to place a pod in the public space and go from there for my common types. I haven't tried that yet. 

I cannot find useful `private specs repo`  steps, but I would think if that were to work - referencing private specs on my own computer would be supported too (as I reference them from a Podfile).

thanks.

**Comment by pacu** [comment]
@konacurrents if you have a private artifact that needs to rely on private pods, that's totally doable by running your own podspec server which can actually be a git repo somewhere. I hadn't set that up myself but I worked on a big company where we had a huge app modularized on private pods (this was prior to SPM even existing) and the different modules were cocoapods in private repos that were "published" to a private `pod repo` that you had to set up on your developer environment to use a source. 

d

**Comment by konacurrents** [comment]
It would be nice to see actual instructions for doing this. I've tried referencing my GitHub but I can't find `podspec` syntax to reference other than known pods and their versions.

**Comment by DominatorVbN** [comment]
> when you run pod lib lint you should config --sources=https://github.com/private-repo-name.git,and same when pod repo push ,like this pod lib lint name.podspec --sources=https://github.com/CocoaPods/Specs.git,192.168.0.100:Plutoy/Specs.git here is the article http://www.pluto-y.com/cocoapod-private-pods-and-module-manager/

This should be added to: https://guides.cocoapods.org/making/private-cocoapods.html

In case this is too open source, and i can make an PR to add the details i would be happy to do so, just point me to the files. 😀

**Comment by oleh-petrenko** [comment]
Hello!
If I'm not mistaken my task is similar to yours. For my project on Flutter I had to use private XCFramework. I'd already had all needed .h and .mm files to import in bridge header. Using plugin as an approach turned out to be a challenging task indeed. I decided to do it in this way: 

- do not use plugin
- open your Xcode project from Flutter project (not from plugin)
- add all needed files in Swift/Objective-C. In my case it was: .h, .mm, XCFrmework
- make sure your Build Settings in Xcode set up correctly (IMPORTANT!)
- then create method channel in Xcode. The thing is Xcode project from Flutter project is already to use method channel. You do not need to edit anything in podfile. Example:

guard let controller = window?.rootViewController as? FlutterViewController ...
var methodChannel: FlutterMethodChannel? = FlutterMethodChannel(name: "test", binaryMessenger: controller.binaryMessenger)
methodChannel?.setMethodCallHandler({ [weak self] (call: FlutterMethodCall, result: @escaping FlutterResult) -> Void in ...))}
 methodChannel?.setMethodCallHandler({ [weak self] (call: FlutterMethodCall, result: @escaping FlutterResult) -> Void in ...
 
 - next step in Flutter part. Make same creation of Method channel with exactly same name ("test")
 - send message from Flutter to Swift and call result with outcome of your operation in Swift
 
 In my situation it worked nice
 
 Good luck!

**Comment by Binozo** [comment]
This is such an embarrassing issue.

Just like the developers in #3276 I am facing the same issue: I need a way to set the dependency/source to a local directory.

 In my case I am contributing to a flutter plugin for the onnxruntime. The official onnxruntime publisher (Microsoft) didn't update their CocoaPods in a while, that's a problem because I need a more recent version of that package.

So I built those myself and now need a way to reference them locally as a dependency - which doesn't work (because of ideology?).

I already did this using AARs (for Android) for the same reason - no problem on the Android side 👍 It just works™

`"Ehmmm.. just add a private pod repo 🤓☝️"` - if it only would be that easy. Let me break it down:

1. I need to host a git repo for this **one** (!!!) package
2. I need to convince the flutter plugin owner to either create such a repo himself (for that one package) or allow me to link the plugin with my own repo created for that purpose. I think in both cases the owner will tell me that I am crazy.
3. Link that podspec file with my repo and pray that it works 🙏

Is this the intended way?
